### ISSUE TITLE
The Order of @TaskAction

### ISSUE THREAD

**Issue by 2BAB** [gradle/gradle#8118]
In a nutshell, the annotation `@TaskAction` doesn't guarantee the order of actions or indicate that it should be used only one time per task.

### Expected Behavior

Some suggestions list below:

1. Provide a default sorting for those action methods that added with `@TaskAction`, ex: Alphabetical Order.
2. Optimize the API doc of `@TaskAction` and relevant user guides, to explicitly indicate what is the best practice for using this annotation.
    - https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskAction.html
    - https://docs.gradle.org/current/userguide/custom_tasks.html
    - https://guides.gradle.org/writing-gradle-tasks/
3. Provide an annotation like `@TaskAction(int priority)` to support more flexible usage.

### Current Behavior

Check the class `DefaultTaskClassInfoStore` and method `createTaskClassInfo(Class<? extends Task> type)`, we can find how `@TaskAction` methods added to the actions list. 

```  java
public class DefaultTaskClassInfoStore implements TaskClassInfoStore {
      ...
      private TaskClassInfo createTaskClassInfo(Class<? extends Task> type) {
        boolean cacheable = type.isAnnotationPresent(CacheableTask.class);
        boolean incremental = false;
        Map<String, Class<?>> processedMethods = Maps.newHashMap();
        ImmutableList.Builder<TaskActionFactory> taskActionFactoriesBuilder = ImmutableList.builder();
        for (Class current = type; current != null; current = current.getSuperclass()) {
            for (Method method : current.getDeclaredMethods()) {
                
                // The order of actions depends on what `getDeclaredMethods()` returns here,
                // however, JDK doesn't guarantee the order of the returning array from this method:
                //
                // > The elements in the returned array are not sorted and are not in any particular order.
                //
                // Thus if you have multiple `@TaskAction` methods, it will be very confusing.

                TaskActionFactory taskActionFactory = createTaskAction(type, method, processedMethods);
                if (taskActionFactory == null) {
                    continue;
                }
                if (taskActionFactory instanceof IncrementalTaskActionFactory) {
                    if (incremental) {
                        throw new GradleException(String.format("Cannot have multiple @TaskAction methods accepting an %s parameter.", IncrementalTaskInputs.class.getSimpleName()));
                    }
                    incremental = true;
                }
                taskActionFactoriesBuilder.add(taskActionFactory);
            }
        }

        return new TaskClassInfo(incremental, taskActionFactoriesBuilder.build(), cacheable);
      }

```

### Context

I created an artifact uploading task with multiple assets, I just added more than one `@TaskAction` on several methods to complete the demonstrate. When I tried to improve it (to define the order and do error handler, etc.), I found that I can not use them anymore. I quite understood I should use "doLast{}" or annotate only one method and place all logic inside without using multiple `@TaskAction`. But I think it was because I'm writing something in `buildSrc` directory that I don't want to reuse them in any other project, it makes me be a lazy boy to do all things together in the task. That's why I had an idea to take more than one `@TaskAction`.

I don't know if Gradle should improve such this small issue, although it is quite an essential API, it may let people feel confused when they first encounter Gradle. 

### Steps to Reproduce (for bugs)

N/A.

### Your Environment

Gradle 5.0 (But I suppose it exists since the beginning)

**Comment by grv87** [comment]
+1 for this.

Documentation on [incremental tasks](https://docs.gradle.org/5.0/userguide/custom_tasks.html#incremental_tasks):
> A task may only contain a single incremental task action

makes me to wonder: is it legal to have several non-incremental task actions, and how would they be ordered?

**Comment by wolfs** [comment]
We should deprecate having more than one @TaskAction on a task class and remove it in the end.

**Comment by lptr** [comment]
Let's keep this open.

**Comment by tg-freigmbh** [comment]
Yep ran into this as well. We accidentally had multiple taskActions in a class and it appears the order of execution changed between 8.9 and 8.12.1. Would be nice if a warning/error would be printed somewhere
